src/conic_form.jl

```diff
@@ -0,0 +1,250 @@
+mutable struct ModelData{T}
+    m::Int # Number of rows/constraints
+    n::Int # Number of cols/variables
+    I::Vector{Int} # List of rows
+    J::Vector{Int} # List of cols
+    V::Vector{T} # List of coefficients
+    b::Vector{T} # constants
+    objective_constant::T # The objective is min c'x + objective_constant
+    c::Vector{T}
+end
+
+mutable struct ConicData{T}
+    f::Int # number of linear equality constraints
+    l::Int # length of LP conic
+    q::Int # length of SOC conic
+    qa::Vector{Int} # array of second-order conic constraints
+    s::Int # length of SD conic
+    sa::Vector{Int} # array of semi-definite constraints
+    ep::Int # number of primal exponential conic triples
+    ed::Int # number of dual exponential conic triples
+    p::Vector{Float64} # array of power conic params
+    nrows::Dict{Int, Int} # The number of rows of each vector sets, this is used by `constrrows` to recover the number of rows used by a constraint when getting `ConstraintPrimal` or `ConstraintDual`
+    data::Union{Nothing, ModelData{T}}
+
+    function ConicData{T}() where T
+        return new{T}(0, 0, 0, Int[], 0, Int[], 0, 0, Float64[], Dict{Int, Int}(), nothing)
+    end
+end
+
+# The ordering of CONES matches SCS
+const CONES = Dict(
+    MOI.Zeros => 1,
+    MOI.Nonnegatives => 2,
+    MOI.SecondOrderCone => 3,
+    MOI.PositiveSemidefiniteConeTriangle => 4,
+    MOI.ExponentialCone => 5, 
+    MOI.DualExponentialCone => 6
+)
+
+const CI = MOI.ConstraintIndex
+const VI = MOI.VariableIndex
+
+cons_offset(conic::MOI.Zeros) = conic.dimension
+cons_offset(conic::MOI.Nonnegatives) = conic.dimension
+cons_offset(conic::MOI.SecondOrderCone) = conic.dimension
+cons_offset(conic::MOI.PositiveSemidefiniteConeTriangle) = Int64((conic.side_dimension*(conic.side_dimension+1))/2)
+
+function restructure_arrays(_s::Array{T}, _y::Array{T}, cones::Array{<: MOI.AbstractVectorSet}) where {T}
+    i=0
+    s = Array{T}[]
+    y = Array{T}[]
+    for conic in cones
+        offset = cons_offset(conic)
+        push!(s, _s[i.+(1:offset)])
+        push!(y, _y[i.+(1:offset)])
+        i += offset
+    end
+    return s, y
+end
+
+# Computes conic dimensions
+function constroffset(conic::ConicData{T}, ci::CI{<:MOI.AbstractFunction, MOI.Zeros}) where T
```

Comment:
A better name would be desirable if possible, `constroffset` gets hard to read.
constraint_offset?

Reply:
yep. editing